modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/rebalance/DistributionZoneRebalanceEngine.java

```diff
@@ -168,30 +138,34 @@ public CompletableFuture<Void> onUpdate(WatchEvent evt) {
 
                     int zoneId = extractZoneId(evt.entryEvent().newEntry().key());
 
-                    DistributionZoneView zoneConfig;
+                    // It is safe to get the latest version of the catalog as we are in the metastore thread.
+                    int catalogVersion = catalogService.latestCatalogVersion();
 
-                    try {
-                        zoneConfig = getZoneById(zonesConfiguration, zoneId).value();
-                    } catch (DistributionZoneNotFoundException e) {
-                        //The zone was removed.
-                        return completedFuture(null);
-                    }
+                    // TODO: IGNITE-20114 Should be get from the catalog directly
+                    // CatalogZoneDescriptor zoneDescriptor = catalogService.zone(zoneId, catalogVersion);
+
+                    String zoneName = distributionZoneManager.getZoneName(zoneId);
+
+                    assert zoneName != null : zoneId;
+
+                    CatalogZoneDescriptor zoneDescriptor = catalogService.zones(catalogVersion).stream()
```

Comment:
Doesn't `zone(String, int)` exist on `CatalogService` to get a zone by name at a version? For getting a table, there is such a method, how about adding it for zones as well?

Reply:
This is a dirty temporary solution, later I think we will not need such a method (which you suggest), so I would not like to add it just like that.